### PR TITLE
Clean up and restructure String bridging a bit

### PR DESCRIPTION
**Pull request by Catfish-Man**
Initial local benchmarking suggests this isn't a speedup, but I'm curious what the full test suite will say.

### CODE REVIEW

**Comment by Catfish-Man** [comment]
@swift-ci please smoke benchmark

File: stdlib/public/core/StringUTF16View.swift [hunk 1]
```diff
@@ -480,7 +483,7 @@ extension String.UTF16View {
   internal func _nativeGetIndex(for offset: Int) -> Index {
     // Trivial and common: start
     if offset == 0 { return startIndex }
-
+
```
Comment (Catfish-Man) [inline stdlib/public/core/StringUTF16View.swift]: I had an ASCII fast path here too but I got it wrong and it was crashing. Need to debug further.

File: stdlib/public/core/StringStorage.swift [hunk 2]
```diff
@@ -12,160 +12,150 @@
...
+      }
+      
+      //Handle the case where both strings were bridged from Swift.
+      //We can't use String.== because it doesn't match NSString semantics.
+      if let nativeOther = other as? _AbstractStringStorage {
+        if count != nativeOther.count {
+          return 0
+        }
+        defer { _fixLifetime(nativeOther) }
+        let otherStart = nativeOther.start
```
Comment (Catfish-Man) [inline stdlib/public/core/StringStorage.swift]: Moving start to the protocol certainly cleaned this up :)

**Comment by Catfish-Man** [comment]
@swift-ci please smoke benchmark

**Comment by Catfish-Man** [comment]
@swift-ci please smoke benchmark

**Comment by milseman** [comment]
@swift-ci please benchmark

**Comment by Catfish-Man** [comment]
@swift-ci please benchmark

**Comment by Catfish-Man** [comment]
@swift-ci please benchmark

**Review by milseman — APPROVED** [review]
LGTM

File: stdlib/public/core/StringBridge.swift [hunk 3]
```diff
@@ -131,6 +131,27 @@ private var kCFStringEncodingUTF8 : _swift_shims_CFStringEncoding {
...
+    switch unsafeBitCast(type(of: str), to: UInt.self) {
+    case _KnownCocoaString.storageType:
+      self = .storage(unsafeBitCast(str, to: _StringStorage.self))
+    case _KnownCocoaString.sharedType:
+      self = .shared(unsafeBitCast(str, to: _SharedStringStorage.self))
+    default:
+      self = .cocoa(str)
+    }
+  }
+}
```
[the remark this answers is not in the document]
Comment (Catfish-Man) [reply stdlib/public/core/StringBridge.swift]: Definitely open to using something better here! I'm not familiar with all the tools in the toolbox yet. What would you suggest for this?

[the remark this answers is not in the document]
Comment (Catfish-Man) [reply stdlib/public/core/StringBridge.swift]: Also the initial version of this hack seems to be a slowdown so we'll see if the one I just pushed looks better (it does locally). If not I'll drop it and keep the rest.

**Comment by Catfish-Man** [comment]
@swift-ci please benchmark

**Comment by Catfish-Man** [comment]
@swift-ci please smoke benchmark

**Comment by Catfish-Man** [comment]
@swift-ci please smoke benchmark

**Comment by Catfish-Man** [comment]
@swift-ci please test and merge

**Comment by Catfish-Man** [comment]
@swift-ci please test and merge

**Comment by Catfish-Man** [comment]
@swift-ci please test and merge

**Comment by Catfish-Man** [comment]
@swift-ci please test and merge

**Comment by Catfish-Man** [comment]
@swift-ci please smoke test

**Comment by Catfish-Man** [comment]
@swift-ci please smoke benchmark

**Comment by Catfish-Man** [comment]
@swift-ci please smoke test

**Comment by Catfish-Man** [comment]
@swift-ci please smoke benchmark

**Comment by Catfish-Man** [comment]
@swift-ci please smoke test and merge

**Comment by Catfish-Man** [comment]
the ABI diffs are due to #SR-9372

**Comment by Catfish-Man** [comment]
@swift-ci please smoke test and merge

**Comment by Catfish-Man** [comment]
@airspeedswift we may want this for Swift 5, not so much because of the perf win (though that's nice), but because it'll make cherry-picking other bridging changes hard if we don't

**Pull request merged**